Implement xtick and ytick rotation modes

kyracho opened the pull request:
Hello, the goal of this PR is to implement two new `rotation_mode` types `'xtick'` and `'ytick'` for the axis labels. 
The idea is that these new rotation_modes automatically set the label alignments to match any `rotation` in $[0,360)$. `'xtick'` aligns the x-axis labels, and 'ytick' aligns the y-axis labels. 

Closes #28951 
### Plots illustrating the proposed `'xtick'` and `'ytick'` rotation modes:

For bottom and left labels:
```python 
import matplotlib.pyplot as plt
import numpy as np

t = np.linspace(0, 2 * np.pi, 1000)
r = np.cos(6 * t)
x_flower = r * np.cos(t)
y_flower = r * np.sin(t)

angles, n_cols = np.linspace(0, 360, 25), 5
fig, axs = plt.subplots(nrows=5, ncols=n_cols, figsize=(30, 30))
axs = axs.flatten()
labels = ['label' for i in range(8)]

for i, angle in enumerate(angles):
    ax = axs[i]
    ax.set_title(f'Rotation: {angle:.0f}°', fontsize=10)

    ax.set_xticks(np.linspace(0, 1.3, 8).tolist())
    ax.set_xticklabels(labels)
    # ax.xaxis.tick_top()
    ax.xaxis.set_tick_params(rotation = angle, rotation_mode='xtick')

    ax.set_yticks(np.linspace(0, 1.3, 8).tolist())
    ax.set_yticklabels(labels)
    # ax.yaxis.tick_right()
    ax.yaxis.set_tick_params(rotation = angle, rotation_mode='ytick')

    ax.set_xlim(0, 1.3)
    ax.set_ylim(0, 1.3)
    ax.plot(x_flower * 0.5 + 0.65, y_flower * 0.5 + 0.65, color='purple')


plt.tight_layout()
plt.savefig('rotated_ticks_figure')
plt.show()
```
![rotated_ticks_figure-4](https://github.com/user-attachments/assets/836159d8-d301-4f10-9ae0-8f4f8394289f)


For top and right labels:
```python 
import matplotlib.pyplot as plt
import numpy as np

t = np.linspace(0, 2 * np.pi, 1000)
r = np.cos(6 * t)
x_flower = r * np.cos(t)
y_flower = r * np.sin(t)

angles, n_cols = np.linspace(0, 360, 25), 5
fig, axs = plt.subplots(nrows=5, ncols=n_cols, figsize=(30, 30))
axs = axs.flatten()
labels = ['label' for i in range(8)]

for i, angle in enumerate(angles):
    ax = axs[i]
    ax.set_title(f'Rotation: {angle:.0f}°', fontsize=10)

    ax.set_xticks(np.linspace(0, 1.3, 8).tolist())
    ax.set_xticklabels(labels)
    ax.xaxis.tick_top()
    ax.xaxis.set_tick_params(rotation = angle, rotation_mode='xtick')

    ax.set_yticks(np.linspace(0, 1.3, 8).tolist())
    ax.set_yticklabels(labels)
    ax.yaxis.tick_right()
    ax.yaxis.set_tick_params(rotation = angle, rotation_mode='ytick')

    ax.set_xlim(0, 1.3)
    ax.set_ylim(0, 1.3)
    ax.plot(x_flower * 0.5 + 0.65, y_flower * 0.5 + 0.65, color='purple')


plt.tight_layout()
plt.savefig('rotated_ticks_figure')
plt.show()
```

![rotated_ticks_figure-3](https://github.com/user-attachments/assets/1d9770b6-8c32-4331-b544-50e199d882b4)

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->
- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

scottshambaugh commented on lib/matplotlib/text.py:
```diff
@@ -1380,6 +1386,30 @@ def set_fontname(self, fontname):
         """
         self.set_fontfamily(fontname)
 
+    def ha_for_angle(self, angle):
```
`angle` here and below should be restricted to [0, 360) deg. Docstring should also call out that `angle` is measured in degrees and not radians

scottshambaugh replied:
I don't think this is being covered yet - the angle input here and the va function should be wrapped to [0, 360). In other words: `angle = angle % 360`

kyracho replied:
Sorry if I didn't mention this earlier! Matplotlib already applies the modulo operation to the input angle before it reaches `_ha_for_angle` and `_va_for_angle`. It applies the modulo operation here:

```python 
 def set_rotation(self, s):
        """
        Set the rotation of the text.

        Parameters
        ----------
        s : float or {'vertical', 'horizontal'}
            The rotation angle in degrees in mathematically positive direction
            (counterclockwise). 'horizontal' equals 0, 'vertical' equals 90.
        """
        if isinstance(s, Real):
            self._rotation = float(s) % 360
        elif cbook._str_equal(s, 'horizontal') or s is None:
            self._rotation = 0.
        elif cbook._str_equal(s, 'vertical'):
            self._rotation = 90.
        else:
            raise ValueError("rotation must be 'vertical', 'horizontal' or "
                             f"a number, not {s}")
        self.stale = True
```

#### Example
Temporarily adding print statements:
```python 
def _ha_for_angle(self, angle, is_tick_top_enabled):
        """
        Determines horizontal alignment ('ha') based on the angle of rotation
        in degrees. Adjusts for is_tick_top_enabled.
        """
        print('angle:', angle)
        if (angle < 5 or 85 <= angle < 105 or 355 <= angle < 360 or
                170 <= angle < 190 or 265 <= angle < 275):
            return 'center'
        elif 5 <= angle < 85 or 190 <= angle < 265:
            return 'left' if is_tick_top_enabled else 'right'
        return 'right' if is_tick_top_enabled else 'left'

    def _va_for_angle(self, angle, is_tick_right_enabled):
        """
        Determines vertical alignment ('va') based on the angle of rotation
        in degrees. Adjusts for is_tick_right_enabled.
        """
        print('angle:', angle)
        if (angle < 5 or 355 <= angle < 360 or 170 <= angle < 190
                or 85 <= angle < 105 or 265 <= angle < 275):
            return 'center'
        elif 190 <= angle < 265 or 5 <= angle < 85:
            return 'baseline' if is_tick_right_enabled else 'top'
        return 'top' if is_tick_right_enabled else 'baseline'
```
Exercising the code path:
```python 
import matplotlib.pyplot as plt
import numpy as np

angles, n_cols = np.linspace(-45, 405, 2), 2
fig, axs = plt.subplots(nrows=1, ncols=n_cols, figsize=(4, 2.5))
axs = axs.flatten()

for i, angle in enumerate(angles):
    ax = axs[i]
    ax.set_title(f'Rotation: {angle:.0f}°', fontsize=10)

    ax.set_xticks([0.5])
    ax.set_xticklabels(['L'])
    ax.xaxis.set_tick_params(rotation=angle, rotation_mode='xtick')

    ax.set_yticks([0.5])
    ax.set_yticklabels(['L'])
    ax.yaxis.set_tick_params(rotation=angle, rotation_mode='ytick')

plt.savefig('rotated_ticks_figure.png')
plt.show()
```
![rotated_ticks_figure](https://github.com/user-attachments/assets/294913ab-1bc3-482d-8e50-45b3413a7c5b)

Terminal:
```
angle: 315.0
angle: 45.0
```
I restricted the angle to [0,360) from [0,360] based on your suggestion. If you want the modulo operations to be added to `_va_for_angle` and `_ha_for_angle` as an additional failsafe, let me know. Thanks!

scottshambaugh commented on lib/matplotlib/text.py:
```diff
@@ -310,7 +310,7 @@ def set_rotation_mode(self, m):
         if m is None:
```
`set_rotation_mode` docstring should be updated with the new modes.

scottshambaugh commented on lib/matplotlib/text.py:
```diff
@@ -1380,6 +1386,30 @@ def set_fontname(self, fontname):
         """
         self.set_fontfamily(fontname)
 
+    def ha_for_angle(self, angle):
```
I think these two new functions are best privatized (prefix the function name with `_`).

scottshambaugh reviewed (commented):
This is looking pretty good I think! I know this is still a draft so apologies if I'm jumping the gun with a review, but please add:
* Image comparison tests exercising the new codepaths
* A "What's new" entry explaining the new feature
* Updates to some of the examples to show this behavior. Good candidate might be https://matplotlib.org/devdocs/gallery/subplots_axes_and_figures/align_labels_demo.html or https://matplotlib.org/devdocs/gallery/images_contours_and_fields/image_annotated_heatmap.html

anntzer commented:
I didn't actually check the implementation, but does this also work with ticks on the right or the top?

kyracho commented:
> This is looking pretty good I think! I know this is still a draft so apologies if I'm jumping the gun with a review, but please add... 

Thanks for the feedback! Added an image comparison test just now. I'll work on the "What's new" entry and the examples once I'm confident that the new changes are OK. 

> I didn't actually check the implementation, but does this also work with ticks on the right or the top?

Thanks! It didn’t handle ticks on the right or top before, but I've updated it to work with those ticks now.

timhoffm commented on lib/matplotlib/text.py:
```diff
@@ -1380,6 +1391,30 @@ def set_fontname(self, fontname):
         """
         self.set_fontfamily(fontname)
 
+    def _ha_for_angle(self, angle, is_tick_top_enabled):
+        """
+        Determines horizontal alignment ('ha') based on the angle of rotation
+        in degrees. Adjusts for is_tick_top_enabled.
+        """
+        if (angle < 5 or 85 <= angle < 105 or 355 <= angle < 360 or
+                170 <= angle < 190 or 265 <= angle < 275):
+            return 'center'
+        elif 5 <= angle < 85 or 190 <= angle < 265:
+            return 'left' if is_tick_top_enabled else 'right'
+        return 'right' if is_tick_top_enabled else 'left'
```
```suggestion
    def _ha_for_angle(self, angle):
        """
        Determines horizontal alignment ('ha') for rotation_mode "xtick" based on
        the angle of rotation in degrees and the vertical alignment.
        """
        anchor_at_bottom = self.get_verticalalignment() == 'bottom'
        if (angle < 5 or 85 <= angle < 105 or 355 <= angle < 360 or
                170 <= angle < 190 or 265 <= angle < 275):
            return 'center'
        elif 5 <= angle < 85 or 190 <= angle < 265:
            return 'left' if anchor_at_bottom else 'right'
        return 'right' if anchor_at_bottom else 'left'
```

If one was really pedantic, this xtick rotation style is only defined for verticalalignment 'top' and 'bottom'. We could error out in the other cases.

kyracho commented on lib/matplotlib/text.py:
```diff
@@ -453,7 +453,18 @@ def _get_layout(self, renderer):
         valign = self._verticalalignment
 
         rotation_mode = self.get_rotation_mode()
+        ax = getattr(self, 'axes', None)
+        if ax:
+            y_tick_params = ax.yaxis.get_tick_params()
+            is_tick_right_enabled = y_tick_params.get('labelright', False)
+            x_tick_params = ax.xaxis.get_tick_params()
+            is_tick_top_enabled = x_tick_params.get('labeltop', False)
```
Thanks so much for the detailed feedback, @timhoffm. I really appreciate your patience. All of your suggestions make sense, and I’ll implement the changes accordingly.

kyracho commented on lib/matplotlib/axis.py:
```diff
@@ -2095,6 +2110,8 @@ def set_ticklabels(self, labels, *, minor=False, fontdict=None, **kwargs):
         for pos, (loc, tick) in enumerate(zip(locs, ticks)):
             tick.update_position(loc)
             tick_label = formatter(loc, pos)
+            tick.label1.axes = self.axes
+            tick.label2.axes = self.axes
```
I get what you're saying, and I agree with you. I'll remove it during the next force push.

timhoffm commented on lib/matplotlib/text.py:
```diff
@@ -301,16 +301,16 @@ def set_rotation_mode(self, m):
-            to their horizontal and vertical alignments.  If ``"anchor"``, then
-            alignment occurs before rotation. Passing ``None`` will set the rotation
-            mode to ``"default"``.
+            to their horizontal and vertical alignments.  If ``"anchor"``, ``"xtick"``
+            or ``"ytick", then alignment occurs before rotation. Passing ``None`` will
+            set the rotation mode to ``"default"``.
```
```suggestion
            to their horizontal and vertical alignments.  If ``"anchor"``, then
            alignment occurs before rotation. "xtick" and "ytick" adjust the
            horizontal/vertical alignment so that the text is visually pointing
            towards its anchor point. This is primarily used for rotated tick
            labels and positions them nicely towards their ticks. Passing
            ``None`` will set the rotation mode to ``"default"``.
```

timhoffm commented on lib/matplotlib/text.py:
```diff
@@ -1380,6 +1385,32 @@ def set_fontname(self, fontname):
+
+    def _va_for_angle(self, angle):
+        """
+        Determines vertical alignment ('va') based on the angle of rotation
+        in degrees. Adjusts for is_tick_right_enabled.
```
like for `_ha_for_angle`:
```suggestion
        Determines vertical alignment ('va') for rotation_mode "ytick" base on
        the angle of rotation in degrees and the horizontal alignment.
```

timhoffm reviewed (commented):
The fundamental implementation is already correct. The following are merely quality and style improvements.

kyracho commented on galleries/examples/subplots_axes_and_figures/align_labels_demo.py:
```diff
@@ -17,25 +17,19 @@
+                ax.xaxis.tick_top()
+            ax.set_xticks(np.linspace(0, 1000, 5))
+            ax.set_xticklabels([250 * n for n in range(5)])
+            ax.xaxis.set_tick_params(rotation=55, rotation_mode='xtick')
```
TYSM for explaining so clearly, and for your hard work to help me. I really appreciate it! 🩵

A few questions: 

There seems to be an issue with
```python
 ax.tick_params(axis='x', rotation=55, rotation_mode='xtick')
```
It leads to
```
AttributeError: Line2D.set() got an unexpected keyword argument 'ion_mode'
```
Tracing to
```python 
class Tick(martist.Artist):
                   ⋮
      def __init__(.**kwargs):
                   ⋮
            grid_kw = {k[5:]: v for k, v in kwargs.items()}
```

The fix I came up with is
```python 
grid_kw = {k[5:]: v for k, v in kwargs.items() if k != "rotation_mode"}
```
to filter out `rotation_mode` and stopping `ion_mode` from being passed to the `Line2D` object.

I'm not sure if it would be better to apply it separately, to the `__init__` method, though? 

```python 
ax.tick_params(axis='x', rotation=55, rotation_mode='xtick')
```
seems to apply `rotation_mode` to only the first tick on the axis. Is there a delayed creation of the ticks at render time? so the `rotation_mode` isn't being applied to all the labels? 

I've tried using

```python 
ax.set_xticks(ax.get_xticks())
ax.tick_params(axis='x', rotation=55, rotation_mode='xtick')
```
which seems to create the ticks before setting `tick_params`, but I'm not sure if it's a good solution?

timhoffm commented on lib/matplotlib/axis.py:
```diff
@@ -1089,11 +1093,21 @@ def _translate_tick_params(kw, reverse=False):
-                oldkey: kw_.pop(newkey)
-                for oldkey, newkey in keymap.items() if newkey in kw_
-            }
+            kwtrans = {}
+            for oldkey, newkey in keymap.items():
+                if newkey in kw_:
+                    if is_x_axis and newkey == 'label1On':
+                        kwtrans['labelbottom'] = kw_.pop(newkey)
+                    elif is_x_axis and newkey == 'tick1On':
+                        kwtrans['bottom'] = kw_.pop(newkey)
+                    elif is_x_axis and newkey == 'label2On':
+                        kwtrans['labeltop'] = kw_.pop(newkey)
+                    elif is_x_axis and newkey == 'tick2On':
+                        kwtrans['top'] = kw_.pop(newkey)
+                    else:
+                        kwtrans[oldkey] = kw_.pop(newkey)
```
I'd like to have the bugfix separated from the rotation feature. Either as two commits in this PR, or (slightly perferred) as two separate PRs.

timhoffm reviewed (commented):
This is good, apart from mixing the kw translation bugfix into the new feature.

QuLogic commented on lib/matplotlib/tests/test_text.py:
```diff
@@ -1135,3 +1135,58 @@ def test_font_wrap():
              wrap=True)
+
+
+def _test_ha_for_angle():
```
This function is unused.

kyracho replied:
Hi @QuLogic, it seems like making  `_test_ha_for_angle` and `_test_va_for_angle` private functions causes Codecov to ignore them? Would you like me to remove them? 

I ran these locally, and they pass:
```bash
pytest --pyargs /workspaces/matplotlib/lib/matplotlib/tests/test_axis.py::_test_ha_for_angles
pytest --pyargs /workspaces/matplotlib/lib/matplotlib/tests/test_axis.py::_test_va_for_angles
```

QuLogic commented on lib/matplotlib/tests/test_text.py:
```diff
@@ -1135,3 +1135,58 @@ def test_font_wrap():
+        assert alignment in ['center', 'left', 'right']
+
+
+def _test_va_for_angle():
```
This function is unused.

QuLogic commented on doc/users/next_whats_new/xtick_ytick_rotation_modes.rst:
```diff
@@ -0,0 +1,28 @@
+
+A new feature has been added for handling rotation of xtick and ytick
+labels more intuitively. The new `rotation modes <matplotlib.text.Text.set_rotation_mode>`
+"xtick" and "ytick" automatically adjusts the alignment of rotated tick labels,
```
```suggestion
"xtick" and "ytick" automatically adjust the alignment of rotated tick labels,
```

QuLogic commented on doc/users/next_whats_new/xtick_ytick_rotation_modes.rst:
```diff
@@ -0,0 +1,28 @@
+A new feature has been added for handling rotation of xtick and ytick
+labels more intuitively. The new `rotation modes <matplotlib.text.Text.set_rotation_mode>`
+"xtick" and "ytick" automatically adjusts the alignment of rotated tick labels,
+so that the texts point towards their anchor point, i.e. ticks.  This works for
```
```suggestion
so that the text points towards their anchor point, i.e. ticks.  This works for
```

kyracho commented:
Hi @QuLogic , just checking in on this PR; whenever u have the time, I'd really appreciate your feedback. 

Please let me know if there is anything I can improve. 

Thank you!

tacaswell commented:
Thank you @kyracho !  This is a pretty nice quality of life improvement.
